packages/babel-helper-member-expression-to-functions/src/index.js

```diff
@@ -37,9 +86,171 @@ const handle = {
+          continue;
+        }
+        // prevent infinite loop: unreachable if the AST is well-formed
+        throw new Error(
+          `Internal error: unexpected ${startingOptional.node.type}`,
+        );
+      }
+
+      const { scope } = member;
+      const startingProp = startingOptional.isOptionalMemberExpression()
+        ? "object"
+        : "callee";
+      const startingNode = startingOptional.node[startingProp];
+      const baseNeedsMemoised = scope.maybeGenerateMemoised(startingNode);
+      const baseRef = baseNeedsMemoised ?? startingNode;
+
+      // Compute parentIsOptionalCall before `startingOptional` is replaced
+      // as `node` may refer to `startingOptional.node` before replaced.
+      const parentIsOptionalCall = parentPath.isOptionalCallExpression({
+        callee: node,
+      });
+      startingOptional.replaceWith(toNonOptional(startingOptional, baseRef));
+      if (parentIsOptionalCall) {
+        if (parent.optional) {
+          parentPath.replaceWith(this.optionalCall(member, parent.arguments));
+        } else {
+          parentPath.replaceWith(this.call(member, parent.arguments));
+        }
+      } else {
+        member.replaceWith(this.get(member));
+      }
+
+      let regular = member.node;
+      for (let current = member; current !== endPath; ) {
+        const { parentPath } = current;
+        // skip transforming `Foo.#BAR?.call(FOO)`
+        if (parentPath === endPath && parentIsOptionalCall && parent.optional) {
+          regular = parentPath.node;
+          break;
+        }
+        regular = toNonOptional(parentPath, regular);
+        current = parentPath;
+      }
+
+      let context;
+      const endParentPath = endPath.parentPath;
+      if (
+        t.isMemberExpression(regular) &&
+        endParentPath.isOptionalCallExpression({
+          callee: endPath.node,
+          optional: true,
+        })
+      ) {
+        const { object } = regular;
+        context = member.scope.maybeGenerateMemoised(object);
+        if (context) {
+          regular.object = t.assignmentExpression("=", context, object);
+        }
+      }
+
+      endPath.replaceWith(
+        t.conditionalExpression(
+          t.logicalExpression(
+            "||",
+            t.binaryExpression(
+              "===",
+              baseNeedsMemoised
+                ? t.assignmentExpression("=", baseRef, startingNode)
+                : baseRef,
+              t.nullLiteral(),
+            ),
+            t.binaryExpression("===", baseRef, scope.buildUndefinedNode()),
+          ),
+          scope.buildUndefinedNode(),
+          regular,
+        ),
+      );
+
+      if (context) {
+        const endParent = endParentPath.node;
+        endParentPath.replaceWith(
+          t.optionalCallExpression(
+            t.optionalMemberExpression(
+              endParent.callee,
+              t.identifier("call"),
+              endParent.computed,
```

Comment:
It should be always `false` because we are transforming `foo.bar(qux)` to `foo.bar.call(foo.bar, qux)`